lib/aws-record/record/batch_writer.rb

```diff
@@ -0,0 +1,222 @@
+      SUCCESSFUL_SEND_STATE = 'SUCCESSFUL'.freeze
+      ERROR_SEND_STATE = 'ERROR'.freeze
+
+      # DynamoDb item request limit
+      ITEM_REQUEST_LIMIT = 25
+
+      def initialize(model, client, items = [])
+        super(model, client, items)
+        @state = PENDING_STATE
+        @errors = []
+        @error_struct = Struct.new(:error, :data, :unprocessed_items) do
+          def successful?
+            false
+          end
+        end
+      end
+
+      # This method calls
+      # {http://docs.aws.amazon.com/sdkforruby/api/Aws/DynamoDB/Client.html#batch_write_item-instance_method Aws::DynamoDB::Client#batch_write_item}},
+      # on the items stored in this BatchWriter. Note that this method does
+      # not support delete_request, only put_request and that only items
+      # of the same class can be written using this method. The following
+      # failure scenarios are prevented before being sent to DynamoDB:
+      #
+      #   - One or more tables specified in the BatchWriteItem request does not
+      #      exist.
+      #   - You try to perform multiple operations on the same item in the same
+      #      BatchWriteItem request. For example, you cannot put and delete the
+      #      same item in the same BatchWriteItem request.
+      #   - There are more than 25 requests in the batch.
+      #   - Primary key attributes specified on an item in the request do not
+      #      match those in the corresponding table's primary key schema.
+      #
+      # This method will not prevent the following failure scenarios and will
+      # need to be handled separately:
+      #
+      #   - Any individual item in a batch exceeds 400 KB.
+      #   - The total request size exceeds 16 MB.
+      #
+      # If any items are not successfully submitted, a RuntimeError is raised
+      # and the BatchWriter can be inspected to see what the errors were.
+      # A call to #errors will return an array of response objects that contain
+      # the errors. See the documentation for Response at
+      # http://docs.aws.amazon.com/sdkforruby/api/Seahorse/Client/Response.html
+      #
+      # @param [Hash] opts options to pass on to the client call to
+      #   +#batch_item_write+. See the documentation above in the AWS SDK for
+      #   Ruby V2.
+      # @opts retry_count the number of retries to attempt on failed write
+      # @raise RuntimeError if unsuccessful in submitting all items
+      def save!(opts = {})
+        raise "can't submit if writer in #{@state} state." unless pending?
+        validate_items
+        # the -1 is to make it equal the number of retries requested
+        @retry_times = (opts.delete(:retry_count) { |_| 4 } - 1)
+        retry_items = send_chunks(opts)
+        responses = retry_failed_requests(retry_items, opts)
+        analyze_result(responses)
+      end
+
+      # Perform the save! function but capture the error that is raised on
+      # unsuccessful save.
+      def save(opts = {})
```

Comment:
Given that we've altered the `#save` semantics, this shouldn't suppress client exceptions, unless we can identify specific client exceptions to pick out. Catching all RuntimeErrors doesn't seem like a best practice here.